File: csharp/Betfair.ESAClient/Betfair.ESAClient/Cache/OrderMarket.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Betfair.ESASwagger.Model;

namespace Betfair.ESAClient.Cache
{
    /// <summary>
    /// The cached state of the market
    /// </summary>
    public class OrderMarket
    {
        private readonly OrderCache _orderCache;
        private readonly string _marketId;
        private readonly Dictionary<RunnerId, OrderMarketRunner> _marketRunners = new Dictionary<RunnerId, OrderMarketRunner>();
        private OrderMarketSnap _snap;

        public OrderMarket(OrderCache orderCache, string marketId)
        {
            _orderCache = orderCache;
            _marketId = marketId;
        }

        internal void OnOrderMarketChange(OrderMarketChange orderMarketChange)
        {            

            OrderMarketSnap newSnap = new OrderMarketSnap();
            newSnap.MarketId = _marketId;

            //update runners
            if (orderMarketChange.Orc != null)
            {
                //runners changed
                foreach (OrderRunnerChange orderRunnerChange in orderMarketChange.Orc)
                {
                    OnOrderRunnerChange(orderRunnerChange);
                }
            }
            newSnap.OrderMarketRunners = _marketRunners.Values.Select(omr => omr.Snap);

            //update closed
            IsClosed = orderMarketChange.Closed == true;
            newSnap.IsClosed = IsClosed;

            _snap = newSnap;
        }

        private void OnOrderRunnerChange(OrderRunnerChange orderRunnerChange)
        {
            RunnerId rid = new RunnerId(orderRunnerChange.Id, orderRunnerChange.Hc);
            OrderMarketRunner orderMarketRunner;
            if (!_marketRunners.TryGetValue(rid, out orderMarketRunner))
            {
                orderMarketRunner = new OrderMarketRunner(this, rid);
                _marketRunners[rid] = orderMarketRunner;
            }
            //update the runner
            orderMarketRunner.OnOrderRunnerChange(orderRunnerChange);
        }

        public string MarketId
        {
            get
            {
                return _marketId;
            }
        }

        public bool IsClosed { get; private set; }

        /// <summary>
        /// Takes or returns an existing immutable snap of the market.
        /// </summary>
        public OrderMarketSnap Snap
        {
            get
            {
                return _snap;
            }
        }

        public override string ToString()
        {
            return "OrderMarket{" +
                "MarketId=" + MarketId +
                ", Runners=" + String.Join(", ", _marketRunners.Values) +
                "}";
        }

    }
}

File: java/client/src/main/java/com/betfair/esa/client/cache/order/OrderMarket.java
package com.betfair.esa.client.cache.order;

import com.betfair.esa.client.cache.util.OrderMarketRunner;
import com.betfair.esa.client.cache.util.OrderMarketRunnerSnap;
import com.betfair.esa.client.cache.util.OrderMarketSnap;
import com.betfair.esa.client.cache.util.RunnerId;
import com.betfair.esa.swagger.model.OrderMarketChange;
import com.betfair.esa.swagger.model.OrderRunnerChange;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class OrderMarket {

    private final OrderCache orderCache;
    private final String marketId;
    private final Map<RunnerId, OrderMarketRunner> marketRunners = new ConcurrentHashMap<>();
    private OrderMarketSnap orderMarketSnap;
    private boolean isClosed;

    public OrderMarket(OrderCache orderCache, String marketId) {
        this.orderCache = orderCache;
        this.marketId = marketId;
    }

    public void onOrderMarketChange(OrderMarketChange orderMarketChange) {
        OrderMarketSnap newSnap = new OrderMarketSnap();
        newSnap.setMarketId(this.marketId);

        // update runners
        if (orderMarketChange.getOrc() != null) {
            for (OrderRunnerChange orderRunnerChange : orderMarketChange.getOrc()) {
                onOrderRunnerChange(orderRunnerChange);
            }
        }

        List<OrderMarketRunnerSnap> snaps = new ArrayList<>(marketRunners.size());
        for (OrderMarketRunner orderMarketRunner : marketRunners.values()) {
            snaps.add(orderMarketRunner.getOrderMarketRunnerSnap());
        }

        newSnap.setOrderMarketRunners(snaps);

        isClosed = Boolean.TRUE.equals(orderMarketChange.isClosed());
        newSnap.setClosed(isClosed);

        orderMarketSnap = newSnap;
    }

    private void onOrderRunnerChange(OrderRunnerChange orderRunnerChange) {

        RunnerId runnerId = new RunnerId(orderRunnerChange.getId(), orderRunnerChange.getHc());

        OrderMarketRunner orderMarketRunner =
                marketRunners.computeIfAbsent(runnerId, r -> new OrderMarketRunner(this, r));

        // update the runner
        orderMarketRunner.onOrderRunnerChange(orderRunnerChange);
    }

    public boolean isClosed() {
        return isClosed;
    }

    public OrderMarketSnap getOrderMarketSnap() {
        return orderMarketSnap;
    }

    public String getMarketId() {
        return marketId;
    }

    @Override
    public String toString() {
        StringBuilder runnersSb = new StringBuilder(" ");
        for (OrderMarketRunner runner : marketRunners.values()) {
            runnersSb.append(runner).append(" ");
        }

        return "OrderMarket{"
                + "marketRunners="
                + runnersSb
                + ", marketId='"
                + marketId
                + '\''
                + '}';
    }
}
